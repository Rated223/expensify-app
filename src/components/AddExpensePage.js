import React from 'react';
import { connect } from 'react-redux';
import { ExpenseForm } from '.';
import { expensesActions } from '../actions';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <p>Add expense</p>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(expensesActions.addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
