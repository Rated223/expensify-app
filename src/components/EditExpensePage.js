import React from 'react';
import { connect } from 'react-redux';
import { ExpenseForm } from ".";
import { expensesActions } from "../actions";

const EditExpensePage = (props) => {
  return (
    <div>
      <ExpenseForm 
        expense={props.expense}
        onSubmit={(expense) => {
          props.dispatch(expensesActions.editExpense({ id: props.match.params.id, update: expense }))
          props.history.push('/');
        }}
      />
      <button onClick={() => {
          props.dispatch(expensesActions.removeExpense({ id: props.match.params.id }));
          props.history.push('/');
      }}>
        Remove
      </button>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
}

export default connect(mapStateToProps)(EditExpensePage);
