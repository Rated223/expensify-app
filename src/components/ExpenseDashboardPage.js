import React from 'react';
import { ExpenseList } from '.';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;