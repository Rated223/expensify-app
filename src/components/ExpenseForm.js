import 'react-dates/initialize';
import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

const now = moment();
console.log('now', now.format('MMM Do YYYY'));

class ExpenseForm extends React.Component {
  state = {
    description: '',
    amount: '',
    note: '',
    createdAt: moment(),
    calendarFocused: false,
    error: ''
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }

  onNoteChange =  (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  }
  
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }))
    }
  }

  onCalendarFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: +this.state.amount,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note

      })
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            autoFocus
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount} 
            onChange={this.onAmountChange}
          />
          <SingleDatePicker 
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onCalendarFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for yor expense (optional)"
            value={this.state.notes}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add expense</button>
        </form>
      </div>
    )
  }
} 

export default ExpenseForm;