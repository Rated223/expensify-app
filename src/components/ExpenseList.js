import React from "react";
import { connect } from "react-redux";
import { ExpenseListItem, ExpensesTotal } from ".";
import { getVisibleExpenses } from '../selectors';

export const ExpenseList = (props) => (
  <div>
    {
      props.expenses.length === 0 ? (
        <p>No expenses</p>
       ) : (
         <div>
          {props.expenses.map((expense) => {
            return <ExpenseListItem key={expense.id} {...expense} />;
          })}
         <ExpensesTotal />
         </div>
       )
    }
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  }
};

export default connect(mapStateToProps)(ExpenseList);