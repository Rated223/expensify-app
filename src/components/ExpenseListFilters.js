import 'react-dates/initialize';
import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from "react-dates";
import { filtersActions } from '../actions';
import 'react-dates/lib/css/_datepicker.css';

export class ExpenseListFilters extends React.Component {

  state = {
    calendarFocused: null
  }

  onTextChange = (e) => {
    this.props.setTextFilter({ text: e.target.value });
  }

  onSortChange = (e) => {
    if (e.target.value === "date") {
      this.props.setSortByDate();
    } else if (e.target.value === "amount") {
      this.props.setSortByAmount();
    }
  }
  
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate({ startDate });
    this.props.setEndDate({ endDate });
  }

  onCalendarFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }))
  }

  render() {
    return (
      <div>
        <input 
          type="text" 
          value={this.props.filters.text} 
          onChange={this.onTextChange}
        />
        <select 
          value={this.props.filters.sortBy} 
          onChange={this.onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker 
          startDate={this.props.filters.startDate}
          startDateId="start-date-id"
          endDate={this.props.filters.endDate}
          endDateId="end-date-id"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onCalendarFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
    
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: ({ text }) => dispatch(iltersActions.setTextFilter({ text })),
  setSortByDate: () => dispatch(filtersActions.sortByDate()),
  setSortByAmount: () => dispatch(filtersActions.sortByAmount()),
  setStartDate: ({ startDate }) => dispatch(filtersActions.setStartDate({ startDate })),
  setEndDate: ({ endDate }) => dispatch(filtersActions.setEndDate({ endDate }))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);