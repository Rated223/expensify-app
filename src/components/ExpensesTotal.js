import React from "react";
import { connect } from "react-redux";
import numeral from 'numeral';
import { getVisibleExpenses, getExpensesTotal } from '../selectors';


export const ExpensesTotal = (props) => (
  <div>
    <p>Total: {numeral(props.total).format('$0,0.00')}</p>
  </div>
)

const mapStateToProps = (state) => {
  return {
    total: getExpensesTotal(getVisibleExpenses(state.expenses, state.filters))
  }
}

export default connect(mapStateToProps)(ExpensesTotal);