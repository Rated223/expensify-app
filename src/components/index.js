import AddExpensePage from './AddExpensePage';
import EditExpensePage from './EditExpensePage';
import ExpenseDashboardPage from './ExpenseDashboardPage';
import ExpenseList from './ExpenseList';
import Header from './Header';
import HelpPage from './HelpPage';
import NotFoundPage from './NotFoundPage';

export {
  AddExpensePage,
  EditExpensePage,
  ExpenseDashboardPage,
  ExpenseList,
  Header,
  HelpPage,
  NotFoundPage
}