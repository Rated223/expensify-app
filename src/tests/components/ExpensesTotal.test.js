import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesTotal } from '../../components/ExpensesTotal';

test('Should render ExpensesTotal with $0.00', () => {
  const wrapper = shallow(<ExpensesTotal />);
  expect(wrapper).toMatchSnapshot();
});

test('Should render ExpenseTotal with $150.00', () => {
  const wrapper = shallow(<ExpensesTotal total={150} />);
  expect(wrapper).toMatchSnapshot();
})