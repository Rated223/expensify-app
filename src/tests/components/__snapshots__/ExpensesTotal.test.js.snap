// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseTotal with $150.00 1`] = `
<div>
  <p>
    Total: 
    $150.00
  </p>
</div>
`;

exports[`Should render ExpensesTotal with $0.00 1`] = `
<div>
  <p>
    Total: 
    $0.00
  </p>
</div>
`;
